docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,302 @@
+
+```yaml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-kaniko-build
+spec:
+  source:
+    container:
+      image: docker.io/<registry-repository>/bundle:latest
+      local: /<path-to-local>/github.com/shipwright-io/sample-go/docker-build
+      pruneBundle: true # Disable by default
+  strategy:
+    name: kaniko
+    kind: ClusterBuildStrategy
+  output: ...
+```
+
+The above Build definition should ensure that a step in our [TaskSpec](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/reconciler/buildrun/resources/taskrun.go#L59) definition is appended, which will ensure a pruning of the image in the related container registry.
+
+#### As a Shipwright/CLI contributor I want to provide a mechanism to support local source code
+
+This is related to the CLI workflow, that should support the bundle approach and maintain the bundle base image. The implementation should be generic enough for other workflows, like OpenShift Build, IBM Cloud Code Engine, to consume it.
+
+### Implementation Details/Notes/Constraints [optional]
+
+There is a prototype for educational purposes that can be tested if desire, see the related [documentation](https://github.com/qu1queee/cli/blob/qu1queee/crud_cmd/README.md). This uses the following two branches:
+
+- Changes in Shipwright/Build, see [branch](https://github.com/qu1queee/build/tree/qu1queee/local_source)
+- Changes in Shipwright/CLI, see [branch](https://github.com/qu1queee/cli/tree/qu1queee/crud_cmd)
+
+### Risks and Mitigations
+
+- This requires coordination between Build and CLI. A well defined list of issues for both backlogs should be created.
+
+- Using a container registry to push/pull the source code might bring concerns around performance. This EP only covers functional requirements. Non-functional requirements like performance or scalability should be discuss during the proposed implementation, and via further development cycles.
+
+- The approach of bundles should not lock-in the CLI with a single approach. It should be flexible enough, to extend to new approaches in the future.
+
+## Design Details
+
+### Test Plan
+
+- Requires development of the self-extractable base image in Shipwright/CLI.
+- Requires development of the bundles `pkg` in Shipwright/CLI.
+- Requires development of the API extension and the runtime changes in Shipwright/Build.
+- This requires new unit, integration and a new e2e tests with local source code in both CLI and Build.
+
+### Graduation Criteria
+
+Should be part of any release before our v1.0.0
+
+### Upgrade / Downgrade Strategy
+
+Not apply, this should not break anything, is rather an extension of the API.
+
+### Version Skew Strategy
+
+N/A
+
+## Implementation History
+
+N/A
+
+## Drawbacks
+
+None
+
+## Alternatives
+
+- None at the moment. But we can always support different mechanisms for supporting local source code feature.
```

Comment:
Regarding the bundle approach. I personally would mention the alternative that the bundle only contains the sources but is not self-extracting and that instead, an image provided by us (like the one for `git` that Matthias is currently building) would run as part of the BuildRun and download the bundle image and extract it.

Pro: clear separation of business logic (= the code that gets the bundle into /workspace/source) from the bundle. Allows us to easily make fixes in our code without the need to fix images that are already out somewhere.

Con: is imo slower. Running the bundle image allows Kubernetes to pull the image in parallel with all other images from the same pod (the containers for the build strategy like buildpacks), it also allows to use Kubernetes' layer cache once we use incremental bundle updates for subsequent builds. The separate image that downloads the bundle image would always need to download the full image assuming it has no means to cache at a persistent location.

Reply:
> I personally would mention the alternative that the bundle only contains the sources but is not self-extracting and that instead, an image provided by us (like the one for git that Matthias is currently building) would run as part of the BuildRun and download the bundle image and extract it.

I´m wondering if your proposal is the same as in `Proposal 03: Self-Extractable Base Image`. We will have two images always. One is the self-extractable one, the second one will use the first one as the base layer and build the final `bundle one`.  For both layers, we will have two separate packages.